### DISCUSSION TITLE
Workers in Next 13 app directory

### DISCUSSION THREAD

**Discussion by ghost** [vercel/next.js#47059] [Help]
### Summary

My next 13 ./app directory app needs to do some long running computation in a thread which doesn't block the UI so I want to use a worker. I can't for he life of me get it to work. Is this expected or am I, as I suspect, just an idiot?

Issue one (solved): I'd expect 'use client' to be sufficient to ensure Worker would be defined, but I'd expect wrongly. Ok, no problem I'll create the worker as a ref. and explicitly call window.Worker.

Issue two (dealbreaker): The problem I can't solve is actually getting the worker compiled:

`<Script strategy="worker">` isn't supported in the app directory so that definitely won't work. 

If I do 

`workerRef.current = new window.Worker(' ./myWorker.js'`

everything compiles and `typeof(workerRef.current)` returns `object Worker` but it's a blank worker with no methods or properties. (And if I deliberately introduce a syntax error, no error message appears either at compile or run time).

So, looking online, previous Next worker examples do the following

`workerRef.current = new window.Worker(new URL('./myWorker.js', import.meta.url))`

this fails at compile time with a message to the effect that the file can't be found. I've tried every possible way of expressing the path I can think of but I can't find one that works.

Can someone help me out either by confirming that workers aren't supposed, or pointing me towards working code? Many thanks.

Thank you SO MUCH for reading and helping.

**Comment by ghost** [comment]
Here's the full code I'm trying...

```
const MyComponent = () => {
    …

    const workerRef = useRef()
    useEffect(() => {
        workerRef.current = new window.Worker('./myWorker.js')
	//workerRef.current = new window.Worker(new URL(' ./myWorker.js', import.meta.url))
         workerRef.current.onmessage = event => alert(`WebWorker Response => ${event.data}`)
	}, [])

    ....
}

```

**Comment by ghost** [comment]
Not sure what the issue is in my app, but reduced to a minimal case the same code does in deed run. Hmmm.

https://github.com/gjhltn/worker-demo

**Comment by devforey** [comment]
Any chance you got this working already? @gjhltn

**Comment by ghost** [comment]
i think so. as far as i can tell it was pilot error by me.. i should close this issue thanks.

**Comment by gleydson** [comment]
Any progress @gjhltn? I have same issue

**Reply by ghost** [reply]
Yeah. The example repo I posted above works for me.

**Comment by lacymorrow** [comment]
I just had to remove `{ type: "module" }` from the worker import:

```ts
		worker.current ??= new Worker(new URL("./worker.js", import.meta.url));
```
